internal/pkg/earpointer/heap.go

```diff
@@ -0,0 +1,216 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package earpointer
+
+import (
+	"go/types"
+	"strings"
+
+	"golang.org/x/tools/go/ssa"
+)
+
+// An abstract reference is a local (register) or global variable, possibly
+// annotated with some kind of context (call stack, allocation site etc).
+
+// Reference is the base class for local and global references.
+type Reference interface {
+	Type() types.Type // Return the data type of the underlying local or global
+	String() string
+	Node() ssa.Value
+}
+
+// ReferenceContext represents the calling context of a reference.
+// Typically, it contains a stack of call instructions.
+type ReferenceContext []ssa.Instruction
+
+// ReferenceLocal is a heap partition that represents all abstract
+// objects that a register of reference type can point to in a specific context.
+type ReferenceLocal struct {
+	context *ReferenceContext
+	reg     ssa.Value
+}
+
+func (ref ReferenceLocal) Type() types.Type {
+	return ref.reg.Type()
+}
+
+func (ref ReferenceLocal) Node() ssa.Value {
+	return ref.reg
+}
+
+func (ref ReferenceLocal) String() string {
+	reg := ref.reg
+	if reg.Parent() != nil {
+		return ref.context.String() + reg.Parent().Name() + "." + reg.Name()
+	} else {
+		return ref.context.String() + reg.Name()
+	}
+}
+
+// ReferenceGlobal is a heap partition that represents all abstract
+// objects that a global of reference type can point to.
+type ReferenceGlobal struct {
+	global *ssa.Global
+}
+
+func (ref ReferenceGlobal) Type() types.Type {
+	return ref.global.Type()
+}
+
+func (ref ReferenceGlobal) Node() ssa.Value {
+	return ref.global
+}
+
+func (ref ReferenceGlobal) String() string {
+	return ref.global.String()
+}
+
+type InternalKind int
+
+const (
```

Comment:
Why are VALUEOF and FIELD exported if these are "internal" kinds? How does a user use these?

Reply:
Changed them to a more accurate name "synthetic" since they are synthesized rather than being generated from an SSA local variable or global variable.

Reply:
Point of order: they would be exported to other packages within this project, but because the path has `internal/`, would not be exported to users.

Reply:
These are for exploring the heap. When the pointer analysis finishes, it generates a heap containing Global, Local, and Synthetic references. The user of this analysis will need to implement some logic by manipulating these references. So exporting them seems right to me.